win32ss/user/user32/controls/ghost.c

```diff
@@ -0,0 +1,570 @@
+        hProcess = OpenProcess(PROCESS_TERMINATE, FALSE, pid);
+        if (hProcess)
+        {
+            TerminateProcess(hProcess, -1);
+            CloseHandle(hProcess);
+        }
+
+        DestroyWindow(hwnd);
+        ShowWindowAsync(pData->hwndTarget, SW_SHOWNOACTIVATE);
+
+        DestroyWindow(pData->hwndTarget);
+        pData->hwndTarget = NULL;
+    }
+    else
+    {
+        GetWindowRect(hwnd, &rc);
+        SetWindowPos(pData->hwndTarget, NULL, rc.left, rc.top, 0, 0,
+            SWP_NOSIZE | SWP_NOOWNERZORDER | SWP_NOACTIVATE | SWP_NOSENDCHANGING);
+
+        ShowWindowAsync(pData->hwndTarget, SW_SHOWNOACTIVATE);
+        DestroyWindow(hwnd);
+    }
+}
+
+static void
+Ghost_OnDraw(HWND hwnd, HDC hdc)
+{
+    BITMAP bm;
+    HDC hdcMem;
+    GHOST_DATA *pData = Ghost_GetData(hwnd);
+
+    if (!pData || !GetObject(pData->hbm32bpp, sizeof(bm), &bm))
+        return;
+
+    hdcMem = CreateCompatibleDC(hdc);
+    if (hdcMem)
+    {
+        HGDIOBJ hbmOld = SelectObject(hdcMem, pData->hbm32bpp);
+        {
+            BitBlt(hdc, 0, 0, bm.bmWidth, bm.bmHeight, hdcMem, 0, 0, SRCCOPY | CAPTUREBLT);
+        }
+        SelectObject(hdcMem, hbmOld);
+        DeleteDC(hdcMem);
+    }
+}
+
+static void
+Ghost_OnNCPaint(HWND hwnd, HRGN hrgn)
+{
+    HDC hdc;
+
+    // do the default behaivour
+    if (IsWindowUnicode(hwnd))
+        DefWindowProcW(hwnd, WM_NCPAINT, (WPARAM)hrgn, 0);
+    else
+        DefWindowProcA(hwnd, WM_NCPAINT, (WPARAM)hrgn, 0);
+
+    // draw the original image
+    hdc = GetWindowDC(hwnd);
+    if (hdc)
+    {
+        Ghost_OnDraw(hwnd, hdc);
+        ReleaseDC(hwnd, hdc);
+    }
+}
+
+static BOOL
+Ghost_OnEraseBkgnd(HWND hwnd, HDC hdc)
+{
+    // draw the original image
+    Ghost_OnNCPaint(hwnd, NULL);
+    return TRUE;
+}
+
+static void
+Ghost_OnPaint(HWND hwnd)
+{
+    PAINTSTRUCT ps;
+    HDC hdc = BeginPaint(hwnd, &ps);
+    if (hdc)
+    {
+        // don't draw at here
+        EndPaint(hwnd, &ps);
+    }
+}
+
+static void
+Ghost_OnMove(HWND hwnd, int x, int y)
+{
+    RECT rc;
+    HWND hwndTarget = Ghost_GetTarget(hwnd);
+    if (!IsWindowVisible(hwndTarget))
+        return;
+
+    GetWindowRect(hwnd, &rc);
+
+    // move the target
+    SetWindowPos(hwndTarget, NULL, rc.left, rc.top, 0, 0,
+        SWP_NOSIZE | SWP_NOOWNERZORDER | SWP_NOACTIVATE | SWP_NOSENDCHANGING);
+}
+
+static void
+Ghost_OnDestroy(HWND hwnd)
+{
+    KillTimer(hwnd, GHOST_TIMER_ID);
+}
+
+static void
+Ghost_OnNCDestroy(HWND hwnd)
+{
+    // delete the user data
+    GHOST_DATA *pData = Ghost_GetData(hwnd);
+    if (IsWindowUnicode(hwnd))
+        SetWindowLongPtrW(hwnd, GWLP_USERDATA, 0);
+    else
+        SetWindowLongPtrA(hwnd, GWLP_USERDATA, 0);
+    HeapFree(GetProcessHeap(), 0, pData);
+
+#ifdef __REACTOS__
+    NtUserSetWindowFNID(hwnd, FNID_DESTROY);
+#endif
+}
+
+static void
+Ghost_OnClose(HWND hwnd)
+{
+    KillTimer(hwnd, GHOST_TIMER_ID);
+
+    if (IDYES == MessageBoxW(hwnd, L"Terminate app?", L"Not responding",
```

Comment:
Avoid Yoda conditions please. I'd find it most readable if using a variable to store the result.
Also, the strings should likely be localizable?

Reply:
Localization is not done yet.

Reply:
Localized.